src/client/opamSolution.ml

```diff
@@ -824,15 +824,45 @@ let parallel_apply t
 
   (* 2/ Display errors and finalize *)
 
-  OpamSwitchState.Installed_cache.save
-    (OpamPath.Switch.installed_opams_cache t.switch_global.root t.switch)
-    (OpamPackage.Set.fold (fun nv opams ->
-         let opam =
-           OpamSwitchState.opam t nv |>
-           OpamFile.OPAM.with_metadata_dir None
-         in
-         OpamPackage.Map.add nv opam opams)
-        t.installed OpamPackage.Map.empty);
+  let save_installed_cache failed =
+    OpamSwitchState.Installed_cache.save
+      (OpamPath.Switch.installed_opams_cache t.switch_global.root t.switch)
+      (OpamPackage.Set.fold (fun nv opams ->
+           let pkg_failed =
+             List.exists (function
+                 | `Fetch ps -> List.for_all (OpamPackage.equal nv) ps
+                 | `Build p
+                 | `Change (_, _, p)
+                 | `Install p
+                 | `Reinstall p
+                 | `Remove p -> OpamPackage.equal nv p)
```

Comment:
The failure is specific to build and fetch errors, maybe adding a comment have that information. It's permit to explain why we need to have that check on install cache update.

Reply:
I don't really think this part of the code needs a comment but i do agree that the section (filtering as a whole) needed one so i added one. What do you think?

Reply:
:+1: